### PR TITLE
feat(api): add eTag/version to API

### PR DESCRIPTION
**Pull request by leite08**
Ref. metriport/metriport-internal#410

### Dependencies

- Upstream: none
- Downstream: https://github.com/metriport/metriport-internal/pull/470

### Description

BREAKING CHANGE:
- create patient endpoint returns 201 instead of 200

Other changes:
- add eTag optional API property for medical entities' `PUT`
- throw 412 if eTag mismatches the DB version upon update
- rename MAPI access table PK to `id`
- expose interfaces for Models, so functions can rely on pure shapes instead of bloated classes with properties and functions inherited from sequelize's `Model`
- disable log of DB operations on cloud envs
- accept "If-Match" header and "eTag" request body property (both optional)
- disable Express' ETag default middleware
- return medical data ordered by ID
- add type to Facility.data - closes #414
- update SDK to support eTag
- update docs
- migrations self contained (no ref to class' table name)

### Release Plan

- has a DB migration
- nothing special on the infrastructure (secrets, etc.)
- just merge along dependency

### CODE REVIEW

File: api/app/src/command/medical/organization/update-organization.ts [hunk 1]
```diff
@@ -1,23 +1,24 @@
...
+  Partial<Pick<Organization, "version">> &
+  OrganizationData;
+
+export const updateOrganization = async (
+  orgUpdate: OrganizationUpdateCmd
+): Promise<Organization> => {
+  const { id, cxId, version, name, type, location } = orgUpdate;
+
+  const org = await getOrganizationOrFail({ id, cxId });
+  validateVersionForUpdate(org, version);
```
Comment (leite08) [inline api/app/src/command/medical/organization/update-organization.ts:15]: Need to load and call a validation function explicitly, I couldn't make sequelize to validate the version.

File: api/app/src/routes/medical/schemas/organization.ts [hunk 2]
```diff
@@ -1,13 +1,16 @@
...
-export const organizationSchema = z.object({
+export const organizationCreateSchema = z.object({
   name: z.string().min(1),
   type: orgTypeSchema,
   location: addressSchema,
 });
+export type OrganizationCreate = z.infer<typeof organizationCreateSchema>;
 
-export type Organization = z.infer<typeof organizationSchema>;
+export const organizationUpdateSchema = organizationCreateSchema.merge(baseUpdateSchema);
```
Comment (leite08) [inline api/app/src/routes/medical/schemas/organization.ts:15]: Update schema extending from baseUpdateSchema (which defined the `eTag` inbound/payload prop).

File: api/app/src/routes/medical/organization.ts [hunk 3]
```diff
@@ -49,12 +58,19 @@ router.put(
...
   asyncHandler(async (req: Request, res: Response) => {
     const cxId = getCxIdOrFail(req);
-    const orgId = getFromParamsOrFail("id", req);
-    const data = organizationSchema.parse(req.body);
+    const id = getFromParamsOrFail("id", req);
+    const payload = organizationUpdateSchema.parse(req.body);
 
-    const org = await updateOrganization({ id: orgId, cxId, data });
+    const updateCmd: OrganizationUpdateCmd = {
+      ...baseUpdateSchemaToCmd(payload),
```
Comment (leite08) [inline api/app/src/routes/medical/organization.ts:65]: Convert from `eTag` to `version` - this is outstanding to try to remove by having the property on the model called `eTag`, optional, and string.

File: api/app/src/command/medical/organization/update-organization.ts [hunk 4]
```diff
@@ -1,23 +1,24 @@
...
+      type,
+      location,
     },
-    { where: { id, cxId }, returning: true }
-  );
-  if (count < 1) throw new NotFoundError();
-  // TODO #156 Send this to Sentry
-  if (count > 1) console.error(`Updated ${count} Orgs for id ${id} and cxId ${cxId}`);
-  return rows[0];
+  });
```
Comment (leite08) [inline api/app/src/command/medical/organization/update-organization.ts:23]: Gotta switch to this (cleaner) way to update so sequelize automatically increases the version.

File: api/app/src/command/medical/organization/update-organization.ts [hunk 5]
```diff
@@ -1,23 +1,22 @@
...
-      data,
+export type OrganizationUpdateCmd = Pick<Organization, "id" | "cxId" | "eTag"> & OrganizationData;
+
+export const updateOrganization = async (
+  orgUpdate: OrganizationUpdateCmd
+): Promise<Organization> => {
+  const { id, cxId, eTag, name, type, location } = orgUpdate;
+
+  const org = await getOrganizationOrFail({ id, cxId });
+  validateVersionForUpdate(org, eTag);
```
Comment (leite08) [inline api/app/src/command/medical/organization/update-organization.ts:13]: Need to load and call a validation function explicitly, I couldn't make sequelize to validate the version.

File: api/app/src/models/medical/mapi-access.ts [hunk 6]
```diff
@@ -1,18 +1,13 @@
...
 
   static setup: ModelSetup = (sequelize: Sequelize) => {
     MAPIAccess.init(
       {
         ...BaseModel.baseAttributes(),
-        cxId: {
-          type: DataTypes.STRING,
-          primaryKey: true,
-        },
       },
```
Comment (Orta21) [inline api/app/src/models/medical/mapi-access.ts:11]: Why was cxId removed?

Comment (leite08) [reply api/app/src/models/medical/mapi-access.ts:11]: > rename MAPI access table PK to id

Comment (leite08) [reply api/app/src/models/medical/mapi-access.ts:11]: That's so we can reuse code/logic based on IDs. An additional pro of this is we have a standard for PKs.

**Review by Goncharo — APPROVED** [review]
Overall looks great - conditional approval based on addressing the comments left in the docs

File: docs/medical-api/getting-started/quickstart.mdx [hunk 7]
```diff
@@ -192,4 +186,66 @@ See the [Get Document endpoint](/medical-api/api-reference/document/get-document
 That's it! You're now well equiped to use the Metriport API to access and manage your Patients' medical data.
 
 As next steps, we'd recommend scoping out our [Medical API Reference](/medical-api/api-reference/organization/create-organization)
-in further detail to see the other things you can do with the API.
\ No newline at end of file
+in further detail to see the other things you can do with the API.
+
+## Additional Notes
```
[the remark this answers is not in the document]
Comment (leite08) [reply docs/medical-api/getting-started/quickstart.mdx]: Done 👍

File: docs/medical-api/getting-started/quickstart.mdx [hunk 8]
```diff
@@ -192,4 +186,66 @@ See the [Get Document endpoint](/medical-api/api-reference/document/get-document
...
+
+These are additional information that might help you get the most out of our Medical API.
+
+The base URL for the API is https://api.metriport.com
+
+The URL for the Sandbox is https://api.sandbox.metriport.com (more about Sandbox [here](#sandbox-mode))
+
+### SDK - Software Development Kit
+
+We've developed a SDK to help you get access to our API without the HTTP plumbing. It's currently only available for
```
[the remark this answers is not in the document]
Comment (leite08) [reply docs/medical-api/getting-started/quickstart.mdx]: Done 👍

File: docs/medical-api/getting-started/quickstart.mdx [hunk 9]
```diff
@@ -192,4 +186,66 @@ See the [Get Document endpoint](/medical-api/api-reference/document/get-document
...
+Node-based applications ([let us know](https://www.metriport.com/contact-us) what other languages you'd like to see it available on):
+
+<Snippet file="medical-api-init.mdx" />
+
+The SDK is available as a NPM package and you can access it [here](https://www.npmjs.com/package/@metriport/api).
```
[the remark this answers is not in the document]
Comment (leite08) [reply docs/medical-api/getting-started/quickstart.mdx]: Done 👍

**Pull request merged**